lib/llm/src/block_manager/distributed/leader.rs

```diff
@@ -296,133 +144,59 @@ impl KvbmLeader {
         let state = self.state.clone();
         let world_size = self.config.world_size;
         let timeout = self.config.leader_init_timeout_secs;
+        let host_cfg = self.config.host_blocks_config.clone();
+        let disk_cfg = self.config.disk_blocks_config.clone();
+
+        // capture num_device_blocks so we can set it inside the closure
+        let num_device_blocks_cell = state.num_device_blocks.clone();
+        let num_host_blocks_cell = state.num_host_blocks.clone();
+        let num_disk_blocks_cell = state.num_disk_blocks.clone();
 
         tokio::spawn(async move {
-            let res = ZmqActiveMessageLeader::new(
+            let res = ZmqActiveMessageLeader::new_with_handshake(
                 leader_sockets,
                 world_size,
                 std::time::Duration::from_secs(timeout),
-                cancel,
+                cancel.clone(),
+                move |workers: &[WorkerMetadata]| -> LeaderMetadata {
+                    // Record device blocks: min across workers
+                    if let Some(min_dev) = workers.iter().map(|w| w.num_device_blocks).min() {
+                        num_device_blocks_cell.store(min_dev, Ordering::Release);
```

Comment:
What's the point of calculating this out if we don't store it in the leader metadata and this is a local variable to the function?